### PR TITLE
Treat `{ new Foo() }` as `SynExpr.ObjExpr`

### PR DESCRIPTION
**Pull request by edgarfgp**
## Description

I propose we unify ` { new Foo }`,  `{ new Foo() }` and `{ New Foo with ... }` as `SynExpr.ObjExpr`

### Before

[fantomas tools](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWTBioDuYxVqwgBfADoAnMcknTufAUIAUASjDipMzfNj9BrVWF6IALgAswAW1iWARrAlhzRAHQAhKhMMBeMIZF4BJCwAB4ADlTI0BQmEgCusCAiQA)

```fsharp
{ new Foo  }   // Parsed as SynExpr.ObjExpr

{ new Foo() }  // Parsed as SynExpr.ComputationExpr

{ new Foo() with member this.Bar() = () }  // Parsed as SynExpr.ObjExpr
```

### After

`{ new Foo() }` will still be parsed as `SynExpr.ComputationExpr` but instead we will re-write this to` SynExpr.ObjExpr` e.g.

```fsharp
{ new Foo  }   // Parsed as SynExpr.ObjExpr

{ new Foo() }  // Still Parsed as SynExpr.ComputationExpr 

{ new Foo() with member this.Bar() = () }  // Parsed as SynExpr.ObjExpr

// We are re-writing to  SynExpr.ObjExpr  If a SynExpr.ComputationExpr body consists of a single SynExpr.New, we should convert it to a SynExpr.ObjExpr/
match .. with
 | SynExpr.ComputationExpr(false, SynExpr.New(_, targetType, expr, m), _) ->       
    false, SynExpr.ObjExpr(targetType, Some(expr, None), None, [], [], [], m, rhsExpr.Range), overallTy, overallTy
```

### Motivation
- Better error reporting
- Consistent compiler analysis.
- Prior work for https://github.com/dotnet/fsharp/pull/17387

```fsharp
[<AbstractClass>]
type Foo() = class end

let foo = { new Foo() }
```

The above snippet reports 3 different errors.

- Invalid object expression. Objects without overrides or interfaces should use the expression form 'new Type(args)' without braces.
- Invalid record, sequence or computation expression. Sequence expressions should be of the form 'seq { ... }'
- Instances of this type cannot be created since it has been marked abstract or not all methods have been given implementations. Consider using an object expression '{ new ... with ... }' instead.

After this PR it will only report 1 error.

- Invalid object expression. Objects without overrides or interfaces should use the expression form 'new Type(args)' without braces.

## Checklist

- [x] Test cases added
- [x] Release notes entry updated

### CODE REVIEW

File: src/Compiler/Checking/CheckComputationExpressions.fs [hunk 1]
```diff
@@ -3454,13 +3454,6 @@ let TcSequenceExpressionEntry (cenv: cenv) env (overallTy: OverallTy) tpenv (has
         let validateObjectSequenceOrRecordExpression = not implicitYieldEnabled
 
         match comp with
-        | SynExpr.New _ ->
```
Comment (edgarfgp) [inline src/Compiler/Checking/CheckComputationExpressions.fs:3457]: As we are rewriting `{ new Foo() }` as `SynExpr.ObjExpr` we no longer need this ad-hoc check here.

**Comment by psfinaki** [comment]
Naive question, how hard would it be to fix the problem in the parser, as in - stop parsing `{ new Foo() }` as `ComputationExpr`? Or is it just about avoiding breaking changes here?

File: src/Compiler/Checking/CheckExpressions.fs [hunk 2]
```diff
@@ -10784,6 +10791,14 @@ and TcNormalizedBinding declKind (cenv: cenv) env tpenv overallTy safeThisValOpt
         let isFixed, rhsExpr, overallPatTy, overallExprTy =
             match rhsExpr with
             | SynExpr.Fixed (e, _) -> true, e, NewInferenceType g, overallTy
+            // Currently { new Foo()  } is parsed as a SynExpr.ComputationExpr, but it should be a SynExpr.ObjExpr
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs]: Yeah it is confusing. Updated the comment.

**Comment by edgarfgp** [comment]
@vzarytovskii Do we also need to update the 9.0 release notes ?

**Comment by vzarytovskii** [comment]
> @vzarytovskii Do we also need to update the 9.0 release notes ?

only 9.0.1xx, not 8.0.xxx as I locked the repo for exactly this reason - bumping version and changing a bunch of defaults, so we can set up insertions in net9 only and bump TFM and language version

**Comment by edgarfgp** [comment]
Would love some feedback. Thanks

**Comment by psfinaki** [comment]
@edgarfgp we've discussed this in the team briefly recently - the motivation of the PR is good here, we need to think if there is any better approach. This one is reasonable, just not traditional. If we cannot think of anything better, we can probably get this in like it is now.

I understand this is a prior work and we appreciate you doing things in small chunks! That said, currently we are directing a lot of our time and energy into reviewing the [nullness PR](https://github.com/dotnet/fsharp/pull/15181) - we want to merge it ASAP to properly battletest it for .NET 9. You will help us a lot if you take a glance there also, since you've touched some of the code areas in your work :)

**Comment by edgarfgp** [comment]
> @edgarfgp we've discussed this in the team briefly recently - the motivation of the PR is good here, we need to think if there is any better approach. This one is reasonable, just not traditional. If we cannot think of anything better, we can probably get this in like it is now.
> 
> I understand this is a prior work and we appreciate you doing things in small chunks! That said, currently we are directing a lot of our time and energy into reviewing the [nullness PR](https://github.com/dotnet/fsharp/pull/15181) - we want to merge it ASAP to properly battletest it for .NET 9. You will help us a lot if you take a glance there also, since you've touched some of the code areas in your work :)

Thanks @psfinaki. My approach here is similar to what @brianrourkeboll took in https://github.com/dotnet/fsharp/pull/17352

**Comment by brianrourkeboll** [comment]
> > @edgarfgp we've discussed this in the team briefly recently - the motivation of the PR is good here, we need to think if there is any better approach. This one is reasonable, just not traditional. If we cannot think of anything better, we can probably get this in like it is now.
> > I understand this is a prior work and we appreciate you doing things in small chunks! That said, currently we are directing a lot of our time and energy into reviewing the [nullness PR](https://github.com/dotnet/fsharp/pull/15181) - we want to merge it ASAP to properly battletest it for .NET 9. You will help us a lot if you take a glance there also, since you've touched some of the code areas in your work :)
> 
> Thanks @psfinaki. My approach here is similar to what @brianrourkeboll took in #17352

Right. In fact, if we updated the parser to treat `{ new Foo() }` as `SynExpr.ObjExpr` instead of `SynExpr.ComputationExpr`, we'd still need to rewrite _that_ to `SynExpr.ComputationExpr` in a scenario like `builder { new Foo() }` anyway. So I'm not sure there's an easy way to avoid AST adjustments during typechecking altogether.

**Comment by edgarfgp** [comment]
This is a prior work to implement a lang suggestion see PR description. Let me know if this PR is clear or you need something else for me to address.

**Comment by psfinaki** [comment]
Hey @edgarfgp, sorry this takes time, we were busy with nullness PR for a while. As for this PR, @T-Gro wanted to take a deep look but is off now for a few days. We keep track of this.

I personally now cannot see any better way to do what you're trying to achieve and @brianrourkeboll's argumentation seems convincing to me. But please allow us some more time here.

**Comment by edgarfgp** [comment]
Patience is not passive, on the contrary, it is concentrated strength. [Bruce Lee](https://wisdomquotes.com/bruce-lee-quotes/)

File: src/Compiler/Checking/CheckExpressions.fs [hunk 3]
```diff
@@ -10891,6 +10898,14 @@ and TcNormalizedBinding declKind (cenv: cenv) env tpenv overallTy safeThisValOpt
...
             match rhsExpr with
             | SynExpr.Fixed (e, _) -> true, e, NewInferenceType g, overallTy
+            // { new Foo()  } is parsed as a SynExpr.ComputationExpr.(See pars.fsy `objExpr` rule).
+            // If a SynExpr.ComputationExpr body consists of a single SynExpr.New, and it's not the argument of a computation expression builder type.
+            // Then we should treat it as a SynExpr.ObjExpr and make it consistent with the other object expressions. e.g.
+            // { new Foo } -> SynExpr.ObjExpr
+            // { new Foo() } -> SynExpr.ObjExpr
+            // { New Foo with ... } -> SynExpr.ObjExpr
+            | SynExpr.ComputationExpr(false, SynExpr.New(_, targetType, expr, m), _) ->        
+                false, SynExpr.ObjExpr(targetType, Some(expr, None), None, [], [], [], m, rhsExpr.Range), overallTy, overallTy
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckExpressions.fs:10908]: My concern here is that we don't **usually** do such rewrites, so it feels a little odd. I think that's the question @T-Gro had on triage as well.

File: src/Compiler/Checking/CheckExpressions.fs [hunk 4]
```diff
@@ -7273,14 +7273,21 @@ and TcObjectExpr (cenv: cenv) env tpenv (objTy, realObjTy, argopt, binds, extraI
         overridesAndVirts |> List.iter (fun (m, implTy, dispatchSlots, dispatchSlotsKeyed, availPriorOverrides, overrides) ->
             let overrideSpecs = overrides |> List.map fst
             let hasStaticMembers = dispatchSlots |> List.exists (fun reqdSlot -> not reqdSlot.MethodInfo.IsInstance)
+            let isOverallTyAbstract =
+                match tryTcrefOfAppTy g objTy with
+                | ValueNone -> false
+                | ValueSome tcref -> HasFSharpAttribute g g.attrib_AbstractClassAttribute tcref.Attribs
+                
+            if overrideSpecs.IsEmpty && not (isInterfaceTy g objTy) then
+                errorR (Error(FSComp.SR.tcInvalidObjectExpressionSyntaxForm (), mWholeExpr))
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckExpressions.fs:7282]: Can you please briefly comment on why is this change here necessary if we only want to rewrite an expression? Is it because we now checking one additional "form" of object expression?

Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs:7282]: @vzarytovskii  See [comment](https://github.com/dotnet/fsharp/pull/17388/files#r1667673941), basically we want to keep the same error reporting.
Added tests to show this

File: src/Compiler/Checking/CheckExpressions.fs (hunk 3 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs:10908]: See [comment](https://github.com/dotnet/fsharp/pull/17388#issuecomment-2220570369) but one way of the other we will need to do some AST rewriting to improve this analysis.

Comment (edgarfgp) [reply src/Compiler/Checking/CheckExpressions.fs:10908]: Checking for `new` inside of SynExpr.ComputationExpr` seems adhoc and that is why I’m proposing this change

**Review by psfinaki — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]
